File: ad-sdk-modern/src/main/java/com/smart/android/adsdk/internal/Hq008AdReporter.java
package com.smart.android.adsdk.internal;

import android.content.Context;
import android.os.Build;
import com.google.gson.Gson;
import com.smart.android.adsdk.AdError;
import com.smart.android.adsdk.AdResult;
import com.smart.android.adsdk.AdResultStatus;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import okhttp3.Call;
import okhttp3.MediaType;
import okhttp3.OkHttpClient;
import okhttp3.Request;
import okhttp3.RequestBody;

final class Hq008AdReporter {
    private static final MediaType JSON = MediaType.parse("application/json; charset=utf-8");
    private static final String EVENT_PROGRESS = "AD_PROGRESS";
    private static final String EVENT_COMPLETED = "AD_COMPLETED";
    private static final String EVENT_ERROR = "AD_ERROR";

    private final Context context;
    private final OkHttpClient okHttpClient;
    private final Gson gson;
    private final String channelId;
    private final String reportUrl;
    private final String consentLogUrl;
    private final boolean enabled;

    Hq008AdReporter(
        Context context,
        OkHttpClient okHttpClient,
        Gson gson,
        String channelId,
        String apiBaseUrl
    ) {
        this.context = context;
        this.okHttpClient = okHttpClient;
        this.gson = gson;
        this.channelId = channelId;
        String baseUrl = apiBaseUrl.endsWith("/") ? apiBaseUrl : apiBaseUrl + "/";
        this.reportUrl = baseUrl + "api/v2/ad/report";
        this.consentLogUrl = baseUrl + "api/v2/ad/consent-log-report";
        this.enabled = true;
    }

    Hq008AdReporter() {
        this.context = null;
        this.okHttpClient = null;
        this.gson = null;
        this.channelId = "";
        this.reportUrl = "";
        this.consentLogUrl = "";
        this.enabled = false;
    }

    void requested(String requestId, long createdAtMs, int width, int height, Map<String, Object> timing) {
        Map<String, Object> diagnostics = baseDiagnostics(createdAtMs, timing);
        diagnostics.put("containerWidth", width);
        diagnostics.put("containerHeight", height);
        diagnostics.put("sdk", "ad-sdk-modern");
        report(requestId, EVENT_PROGRESS, "REQUESTED", diagnostics);
    }

    void loaded(String requestId, long createdAtMs, Map<String, Object> timing) {
        report(requestId, EVENT_PROGRESS, "LOADED", baseDiagnostics(createdAtMs, timing));
    }

    void started(String requestId, long createdAtMs, Map<String, Object> timing) {
        report(requestId, EVENT_PROGRESS, "STARTED", baseDiagnostics(createdAtMs, timing));
    }

    void finished(String requestId, long createdAtMs, AdResult result, Map<String, Object> timing) {
        Map<String, Object> diagnostics = baseDiagnostics(createdAtMs, timing);
        diagnostics.put("status", result.getStatus().name());
        diagnostics.put("reason", result.getReason());
        AdError error = result.getError();
        if (error != null) {
            diagnostics.put("errorSource", error.getSource());
            diagnostics.put("originalCode", error.getOriginalCode());
            diagnostics.put("originalMessage", error.getMessage());
            diagnostics.put("errorStage", error.getStage().name());
            diagnostics.put("responseBody", error.getResponseBody());
        }
        if (result.getStatus() == AdResultStatus.COMPLETED) {
            report(requestId, EVENT_COMPLETED, "COMPLETED", diagnostics);
        } else if (result.getStatus() == AdResultStatus.ERROR) {
            report(
                requestId,
                EVENT_ERROR,
                error == null ? "AD_ERROR" : error.getMessage(),
                diagnostics
            );
        } else if (result.getStatus() == AdResultStatus.SKIPPED) {
            report(requestId, EVENT_ERROR, result.getMessage(), diagnostics);
        } else if (result.getStatus() == AdResultStatus.CANCELLED) {
            report(requestId, EVENT_ERROR, "CANCELLED", diagnostics);
        }
    }

    private Map<String, Object> baseDiagnostics(long createdAtMs, Map<String, Object> timing) {
        Map<String, Object> diagnostics = new LinkedHashMap<>();
        diagnostics.putAll(timing);
        diagnostics.put("createdAtMs", createdAtMs);
        diagnostics.put("sdkEntry", "ima");
        diagnostics.put("deviceModel", Build.MODEL == null ? "" : Build.MODEL);
        diagnostics.put("deviceMake", Build.MANUFACTURER == null ? "" : Build.MANUFACTURER);
        return diagnostics;
    }

    private void report(
        String requestId,
        String eventType,
        String message,
        Map<String, Object> diagnostics
    ) {
        if (!enabled || requestId == null || requestId.trim().isEmpty()) {
            return;
        }
        DeviceInfo deviceInfo = DeviceInfo.collect(context);
        Map<String, Object> body = new LinkedHashMap<>();
        body.put("request_id", requestId);
        body.put("event_type", eventType);
        body.put("uuid", deviceInfo.androidId);
        body.put("channel_id", channelId);
        body.put("ad_version", deviceInfo.versionCode);
        body.put("mac", deviceInfo.mac.isEmpty() ? "00:00:00:00:00:00" : deviceInfo.mac);
        body.put("app_id", deviceInfo.packageName);
        body.put("make", deviceInfo.make);
        body.put("model", deviceInfo.model);
        body.put("message", message == null ? "" : message);
        if (!deviceInfo.localIp.isEmpty()) body.put("local_ip", deviceInfo.localIp);
        if (diagnostics != null) body.put("diagnostic_info", gson.toJson(diagnostics));
        send(reportUrl, body, "requestId=" + requestId + " event=" + eventType);
    }

    void consentLog(String eventType, String message, Map<String, Object> summary) {
        if (!enabled) return;
        String adLog = gson.toJson(summary);
        if (adLog.length() > 256_000) {
            // Drop large response bodies, keeping valid JSON, original code/message and all steps.
            Object error = summary.get("error");
            if (error instanceof Map) ((Map<?, ?>) error).remove("responseBody");
            summary.put("traceCompacted", true);
            summary.put("traceCompactedMode", "without_response_body");
            adLog = gson.toJson(summary);
        }
        // JSON escaping can expand even bounded step messages; keep the payload within the wire limit.
        List<?> steps = (List<?>) summary.get("steps");
        while (adLog.length() > 256_000 && steps != null && steps.size() > 2) {
            steps.remove(1);
            summary.put("stepCount", steps.size());
            summary.put("traceCompacted", true);
            summary.put("traceCompactedMode", "latest_steps");
            adLog = gson.toJson(summary);
        }
        if (adLog.length() > 256_000) {
            // Defensive fallback for any future fields added without a bound.
            summary.remove("steps");
            summary.remove("error");
            summary.remove("diagnostics");
            summary.put("stepCount", 0);
            summary.put("traceCompacted", true);
            summary.put("traceCompactedMode", "terminal_only");
            adLog = gson.toJson(summary);
        }
        DeviceInfo deviceInfo = DeviceInfo.collect(context);
        Map<String, Object> body = new LinkedHashMap<>();
        body.put("channel_id", channelId);
        body.put("mac", deviceInfo.mac.isEmpty() ? "00:00:00:00:00:00" : deviceInfo.mac);
        body.put("ad_version", deviceInfo.versionCode);
        body.put("event_type", eventType);
        body.put("event_message", message);
        body.put("ad_log", adLog);
        send(consentLogUrl, body, "consent-log-report event=" + eventType);
    }

    private void send(String url, Map<String, Object> body, String label) {
        Request request = new Request.Builder()
            .url(url)
            .post(RequestBody.create(JSON, gson.toJson(body)))
            .header("Accept", "application/json")
            .build();
        Call call = okHttpClient.newCall(request);
        call.enqueue(new okhttp3.Callback() {
            @Override
            public void onFailure(Call call, java.io.IOException error) {
                SdkLog.e("AdSdkReport", label + " report failed", error);
            }

            @Override
            public void onResponse(Call call, okhttp3.Response response) {
                SdkLog.i("AdSdkReport", label + " report HTTP=" + response.code());
                response.close();
            }
        });
    }
}

File: app/src/hq008/java/com/smart/android/ad_app/Hq008AdReporter.kt
package com.smart.android.ad_app

import android.os.Build
import android.provider.Settings
import android.util.Log
import com.smart.android.ad_app.bean.EmptyData
import com.google.gson.Gson
import io.github.lib_autorun.ext.getMacAddress
import io.github.lib_autorun.net.NetworkHelper
import io.github.lib_autorun.net.enum.RequestMethod
import java.net.Inet4Address
import java.net.NetworkInterface
import java.util.UUID

internal object Hq008AdReporter {
    private val REPORT_URL = "${BuildConfig.BASE_URL}api/v2/ad/report"
    private const val TAG = "Hq008AdReporter"
    private val gson = Gson()

    object EventType {
        const val AD_PROGRESS = "AD_PROGRESS"
        const val AD_COMPLETED = "AD_COMPLETED"
        const val AD_ERROR = "AD_ERROR"
    }

    object Message {
        const val REQUESTED = "REQUESTED"
        const val LOADED = "LOADED"
        const val STARTED = "STARTED"
        const val COMPLETED = "COMPLETED"
        const val CONTAINER_ERROR = "CONTAINER_ERROR"
    }

    fun newRequestId(): String {
        return "hq008-${System.currentTimeMillis()}-${UUID.randomUUID()}"
    }

    fun reportRequested(
        requestId: String,
        adId: String?,
        hiddenMode: Boolean,
        containerWidth: Int,
        containerHeight: Int,
        extra: Map<String, Any?> = emptyMap()
    ) {
        report(
            requestId = requestId,
            eventType = EventType.AD_PROGRESS,
            message = Message.REQUESTED,
            diagnosticInfo = buildDiagnosticInfo(
                adId = adId,
                hiddenMode = hiddenMode,
                extra = mapOf(
                    "containerWidth" to containerWidth,
                    "containerHeight" to containerHeight
                ) + extra
            )
        )
    }

    fun reportLoaded(
        requestId: String,
        adId: String?,
        hiddenMode: Boolean,
        extra: Map<String, Any?> = emptyMap()
    ) {
        report(
            requestId = requestId,
            eventType = EventType.AD_PROGRESS,
            message = Message.LOADED,
            diagnosticInfo = buildDiagnosticInfo(adId, hiddenMode, extra)
        )
    }

    fun reportStarted(
        requestId: String,
        adId: String?,
        hiddenMode: Boolean,
        startProgress: Double?,
        extra: Map<String, Any?> = emptyMap()
    ) {
        report(
            requestId = requestId,
            eventType = EventType.AD_PROGRESS,
            message = Message.STARTED,
            diagnosticInfo = buildDiagnosticInfo(
                adId = adId,
                hiddenMode = hiddenMode,
                extra = mapOf("startProgress" to startProgress) + extra
            )
        )
    }

    fun reportCompleted(
        requestId: String,
        adId: String?,
        hiddenMode: Boolean,
        extra: Map<String, Any?> = emptyMap()
    ) {
        report(
            requestId = requestId,
            eventType = EventType.AD_COMPLETED,
            message = Message.COMPLETED,
            diagnosticInfo = buildDiagnosticInfo(adId, hiddenMode, extra)
        )
    }

    fun reportError(
        requestId: String,
        adId: String?,
        hiddenMode: Boolean,
        errorCode: Int?,
        errorMessage: String,
        extra: Map<String, Any?> = emptyMap()
    ) {
        report(
            requestId = requestId,
            eventType = EventType.AD_ERROR,
            message = errorMessage,
            diagnosticInfo = buildDiagnosticInfo(
                adId = adId,
                hiddenMode = hiddenMode,
                extra = mapOf("errorCode" to errorCode) + extra
            )
        )
    }

    private fun report(
        requestId: String,
        eventType: String,
        message: String,
        diagnosticInfo: String
    ) {
        val params = linkedMapOf<String, Any>(
            "request_id" to requestId,
            "event_type" to eventType,
            "uuid" to resolveDeviceId(),
            "channel_id" to BuildConfig.CHANNEL,
            "mac" to (safeGetMacAddress()?.takeIf { it.isNotBlank() } ?: "00:00:00:00:00:00"),
            "app_id" to appContext.packageName,
            "make" to Build.MANUFACTURER.orEmpty(),
            "model" to Build.MODEL.orEmpty(),
            "message" to message,
            "diagnostic_info" to diagnosticInfo
        )

        resolveLocalIp()?.let { params["local_ip"] = it }
        val requestJson = gson.toJson(params)

        Log.i(
            TAG,
            "report requestId=$requestId eventType=$eventType message=$message body=$requestJson"
        )

        NetworkHelper.makeRequest<EmptyData>(
            url = REPORT_URL,
            method = RequestMethod.POST,
            params = params,
            isEncryted = false,
            useDomainSwitch = false,
        ) { _, error ->
            if (error != null) {
                Log.e(TAG, "report failed requestId=$requestId eventType=$eventType message=$message error=${error.message}", error)
            } else {
                Log.i(
                    TAG,
                    "report success requestId=$requestId eventType=$eventType message=$message"
                )
            }
        }
    }

    private fun buildDiagnosticInfo(
        adId: String?,
        hiddenMode: Boolean,
        extra: Map<String, Any?> = emptyMap()
    ): String {
        val payload = linkedMapOf<String, Any?>(
            "adId" to adId,
            "hiddenMode" to hiddenMode,
            "sdkVersion" to "2.8.02",
            "deviceModel" to Build.MODEL.orEmpty(),
            "deviceMake" to Build.MANUFACTURER.orEmpty()
        )
        payload.putAll(extra)
        return gson.toJson(payload)
    }

    private fun resolveDeviceId(): String {
        return Settings.Secure.getString(
            appContext.contentResolver,
            Settings.Secure.ANDROID_ID
        ) ?: "unknown_device"
    }

    private fun resolveLocalIp(): String? {
        return runCatching {
            NetworkInterface.getNetworkInterfaces().toList()
                .flatMap { it.inetAddresses.toList() }
                .firstOrNull { !it.isLoopbackAddress && it is Inet4Address }
                ?.hostAddress
        }.getOrNull()
    }

    private fun safeGetMacAddress(): String? {
        return runCatching { getMacAddress() }
            .onFailure { error -> Log.w(TAG, "getMacAddress failed: ${error.message}") }
            .getOrNull()
    }
}
